### PR TITLE
Implement TableUriBuilder and TableSasBuilder.Parse

### LINKED ISSUE #11945: Tables: Implement a method on TableSasBuilder to Parse an existing SAS
# Summary
Given an existing Shared Access Signature, there is currently no way to easily parse it into a format that is simple to interpret. 

# Scope of Work
- Implement a `Parse` method on `TableSasBuilder` that populates the relevant fields defined in the provided SAS string.

# Out of Scope

# Success Criteria
- `Parse` method properly interprets all possible SAS token components.
- Failure conditions for malformed or invalid SAS tokens are handled appropriately.

# Related Issues and References
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/tables/Azure.Data.Tables/src/Sas/TableSasBuilder.cs
#13727

### LINKED ISSUE #13727: Implement TableUriBuilder
...To interact with `TableAccountSasQueryParameters`.

### PR DESCRIPTION
**Pull request by christothes**
closes #11945
closes #13727

### CODE REVIEW

File: sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs [hunk 1]
```diff
@@ -316,6 +316,7 @@ public partial class TableSasBuilder
         public override bool Equals(object obj) { throw null; }
         [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
         public override int GetHashCode() { throw null; }
+        public static Azure.Data.Tables.Sas.TableSasBuilder Parse(System.Uri uri) { throw null; }
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs:319]: Why not ctor?

File: sdk/tables/Azure.Data.Tables/src/Sas/TableUriBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                path.Append('/').Append(Tablename);
+            }
+
+            // Concatenate query parameters
+            var query = new StringBuilder(Query);
+            var sas = Sas?.ToString();
+            if (!string.IsNullOrWhiteSpace(sas))
+            {
+                if (query.Length > 0)
+                { query.Append('&'); }
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/src/Sas/TableUriBuilder.cs:195]: nit: formatting

File: sdk/tables/Azure.Data.Tables/src/TableConnectionString.cs [hunk 3]
```diff
@@ -269,6 +271,84 @@ static Uri CreateUri(string endpoint, string sasToken)
             return false;
         }
 
+        private const string TableNameGroup = "tableName";
+        private static readonly Regex s_tableUriTableNameRegex = new Regex(@$"(Tables\('(?<{TableNameGroup}>\w+)'\))");
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/src/TableConnectionString.cs:275]: add timeout

Comment (pakrym) [reply sdk/tables/Azure.Data.Tables/src/TableConnectionString.cs:275]: Compiled? Would indexOf be simpler?

File: sdk/tables/Azure.Data.Tables/src/UriQueryParamsCollection.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+{
+    internal sealed class UriQueryParamsCollection : Dictionary<string, string>
+    {
+        public UriQueryParamsCollection() : base(StringComparer.OrdinalIgnoreCase) { }
+
+        /// <summary>
+        /// Takes encoded query params string, output decoded params map.
+        /// </summary>
+        /// <param name="encodedQueryParamString"></param>
+		public UriQueryParamsCollection(string encodedQueryParamString)
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/src/UriQueryParamsCollection.cs:20]: nit: tabs

File: sdk/tables/Azure.Data.Tables/src/UriQueryParamsCollection.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            encodedQueryParamString = encodedQueryParamString ?? throw Errors.ArgumentNull(nameof(encodedQueryParamString));
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/src/UriQueryParamsCollection.cs:22]: Arguments.*?

File: sdk/tables/Azure.Data.Tables/src/Sas/TableUriBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            }
+
+            // Use RequestUriBuilder, which has slightly nicer formatting
+            return new RequestUriBuilder
+            {
+                Scheme = Scheme,
+                Host = Host,
+                Port = Port,
+                Path = path.ToString(),
+                Query = query.Length > 0 ? "?" + query.ToString() : null
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/src/Sas/TableUriBuilder.cs:206]: prepend the `"?"` to the string builder and avoid an extra allocation.

**Comment by pakrym** [comment]
TableConnectionString coverage is kind of low.

File: sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs (hunk 1 above)
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs:319]: I went back and forth about this. Perhaps the ctor makes more sense here, but I'm not sure.

File: sdk/tables/Azure.Data.Tables/src/TableConnectionString.cs (hunk 3 above)
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableConnectionString.cs:275]: I meant to make it compiled, thanks for the call out. 

I considered doing an indexOf for `Tables('`, but this seems more straight forward than parsing out that, and the trailing `'`.

File: sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs (hunk 1 above)
Comment (pakrym) [reply sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs:319]: I think TableUriBuilder and TableSasBuilder should be consistent one way or another.

File: sdk/tables/Azure.Data.Tables/src/Sas/TableUriBuilder.cs [hunk 7]
```diff
@@ -192,18 +192,24 @@ private RequestUriBuilder BuildUri()
...
-                { query.Append('&'); }
+                {
+                    query.Append('&');
+                }
                 query.Append(sas);
+                query.Insert(0, "?");
             }
 
             // Use RequestUriBuilder, which has slightly nicer formatting
+            if (query.Length > 0)
```
Comment (pakrym) [inline sdk/tables/Azure.Data.Tables/src/Sas/TableUriBuilder.cs:203]: ?

**Review by pakrym — APPROVED** [review]

File: sdk/tables/Azure.Data.Tables/src/Sas/TableUriBuilder.cs (hunk 7 above)
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/Sas/TableUriBuilder.cs:203]: oops - moved things around and left this by accident.

**Pull request merged**